### DISCUSSION TITLE
Calling block_on from within wasm.

### DISCUSSION THREAD

**Discussion by simonwarchol** [tokio-rs/tokio#5667] [Q&A]
I'm trying to call block_on and capture the result of my async function in wasm but getting runtime errors. Does anyone have suggestions?

let result =  tokio::runtime::Builder::new_current_thread()
        .build()
        .unwrap()
        .block_on(async_fn())

**Comment by Darksonn** [comment]
Tokio's WASM support is very limited. Trying to use something unsupported results in a runtime error.

Without further information, I cannot help you.

**Reply by 9SMTM6** [reply]
Is the single threaded runtime supposed to be usable from the main browser thread on WASM? I think I have the same issue, in whatever way I try to run a single_threaded runtime on the main thread I get an error in the browser, Chrome: `Uncaught RuntimeError: Atomics.wait cannot be called in this context`. 

Note that my wasm build has a bunch of features enabled (+atomics,+bulk-memory,+mutable-globals) as I also run `wasm_thread`. I need to run an async task on the main thread however, since I need access to some APIs that are not available (yet and reliably) in web workers.

It seems however on cursory experiments that task::LocalSet works fine, if I enter an async context via wasm_bindgen_futures. I wasn't sure if that was using the global runtime for anything important, so it took me a lot of experimentation to get there.

**Reply by Darksonn** [reply]
No, when the Tokio runtime becomes idle, it will block. Blocking is *not* allowed on the main browser thread in wasm.

**Reply by 9SMTM6** [reply]
Perhaps I can make a documentation PR that makes these constraints a bit more obvious. The [current](https://docs.rs/tokio/latest/tokio/#wasm-support) documentation talks about a panic, but that runtime error isn't a panic per se, as that error is coming from the browser (to my understanding at least), that threw me off a bit. I thought that you might do explicit detection and cause a panic that way. Note that this distinction might come from the rustc target features I enabled, I think you might use some APIs that panic if you compile without these features. 

And it would make things a lot easier for people not experts in these topics if we explicitly state something to the effect of 

> a runtime won't work on the browser main thread if it ever idles, in webworkers it should be fine. Most applications, e.g. egui, will idle. You can however use task::LocalSet everywhere

**Reply by Darksonn** [reply]
You're welcome to submit a PR, but the note about `LocalSet` seems incorrect. Most likely something else is going on.

**Reply by 9SMTM6** [reply]
In what way is does it seem incorrect? That would be helpful, if it works, but documenting things the wrong way only makes things worse:-).

Could be that it is very limited. My small experiment was adding an async task looking on a channel receive to a localSet, and await that from a `wasm_bindgen_futures::spawn_local`. Then I can successfully trigger the code in that loop from elsewhere in my application. And that worked fine.

**Reply by 9SMTM6** [reply]
I guess I'm just gonna be vague that it seems to work in some contexts. I dont have time to test all setups, but I think what I know should already be helpful. I'm now gonna run this in my little application, if I notice anything else I'll be sure to update it.

**Reply by Darksonn** [reply]
Basically, if you end up in a situation where all tasks are idle, you get that error. If you don't end up in such a situation, you don't. Using `LocalSet` has no effect on this.

**Reply by 9SMTM6** [reply]
> Using LocalSet has no effect on this.

With that you mean that an LocalSet will panic just the same as a runtime, or that a localset, as its no runtime, is not influenced by this?

The way I use the LocalSet I have no (tokio) runtime at all - it seems (I'm certainly not starting one myself or indirectoy via a macro). I was using the LocalSet just as a scheduler of sorts, I guess.

I wasn't sure if LocalSet had any dependencies on the tokio runtime, but so far I did not stumble upon a failure. And it works well enough for me, spawn_local was the only thing I would've been able to use for these tasks anyways, as they're not send nor sync.

Btw, the PR is #6835.

**Reply by Darksonn** [reply]
It sounds like you are using the `LocalSet` as a `FuturesUnordered`.
